File: gfg/POTD/2025-02-08/solution.py

'''
class Node:
    def __init__(self, val):
        self.right = None
        self.data = val
        self.left = None
'''
class Solution:
    def boundaryTraversal(self, root):
        if not root:
            return []

        left_border = []
        right_border = []
        leaf_nodes = []

        def traverse_left_border(node):
            if not node or (not node.left and not node.right):
                return
            left_border.append(node.data)
            if node.left:
                traverse_left_border(node.left)
            else:
                traverse_left_border(node.right)

        def traverse_right_border(node):
            if not node or (not node.left and not node.right):
                return
            if node.right:
                traverse_right_border(node.right)
            else:
                traverse_right_border(node.left)
            right_border.append(node.data)

        def collect_leaf_nodes(node):
            if not node:
                return
            if not node.left and not node.right:
                leaf_nodes.append(node.data)
                return
            collect_leaf_nodes(node.left)
            collect_leaf_nodes(node.right)

        # Add root node
        result = [root.data]

        # Traverse left border
        traverse_left_border(root.left)

        # Collect leaf nodes
        collect_leaf_nodes(root.left)
        collect_leaf_nodes(root.right)

        # Traverse right border
        traverse_right_border(root.right)

        # Combine results
        result.extend(left_border)
        result.extend(leaf_nodes)
        result.extend(right_border)

        return result

File: gfg/POTD/2025-02-08/solution.js
/*
class Node{
    constructor(data){
        this.data = data;
        this.left = null;
        this.right = null;
    }
}
*/

/**
 * @param {Node} root
 * @returns {number[]}
 */

class Solution {
	boundaryTraversal(root) {
		if (!root) return []

		const leftBorder = []
		const rightBorder = []
		const leafNodes = []

		const traverseLeftBorder = (node) => {
			if (!node || (!node.left && !node.right)) return
			leftBorder.push(node.data)
			if (node.left) traverseLeftBorder(node.left)
			else traverseLeftBorder(node.right)
		}

		const traverseRightBorder = (node) => {
			if (!node || (!node.left && !node.right)) return
			if (node.right) traverseRightBorder(node.right)
			else traverseRightBorder(node.left)
			rightBorder.push(node.data)
		}

		const collectLeafNodes = (node) => {
			if (!node) return
			if (!node.left && !node.right) {
				leafNodes.push(node.data)
				return
			}
			collectLeafNodes(node.left)
			collectLeafNodes(node.right)
		}

		// Add root node
		const result = [root.data]

		// Traverse left border
		traverseLeftBorder(root.left)

		// Collect leaf nodes
		collectLeafNodes(root.left)
		collectLeafNodes(root.right)

		// Traverse right border
		traverseRightBorder(root.right)

		// Combine results
		result.push(...leftBorder, ...leafNodes, ...rightBorder)

		return result
	}
}
